### ISSUE TITLE
Assets handling

### ISSUE THREAD

**Issue by Learath2** [ddnet/ddnet#6753]
#6724 moving around images has brought up a longstanding issue with futureproofing assets. As with all issues DDNet there is no way to make everyone happy. However, I think @heinrich5991's idea of a zip file holding all assets and @Jupeyy's idea of a directoryful of assets can both be implemented at the same time. It would allow artists to share a coherent zip of assets and it would allow easy customization of single assets (e.g. just changing the gun).

With a zip or a directory of assets any new assets we add we can just fall back to the default if it doesn't exist in the zip or the directory, so we will never have to think about this when adding/moving assets.

Cons:
- No previews for zips (could be mitigated a bit with the discord bot grabbing the zips and generating a preview)
- Old assets (game.png, hud.png) will need converting. (It should be easy enough to code a tool that does this)

Another related issue is mixing assets. I think this setup would be enough, people that want to add something can easily unzip to a directory and replace the things they don't like. So I think we shouldn't have support for mixing in-game. If we do have to do it, it should be as simple as possible, pick one base set and one overrides set. More than 2 sets with integer priorities would be far too confusing

Related: #3442

**Comment by heinrich5991** [comment]
> I think this setup would be enough, people that want to add something can easily unzip to a directory and replace the things they don't like.

They don't need to unzip for this on Windows, they can simply copy the image into the zip file. On Linux, you can also add files to zip files without extracting, but I couldn't do it with a simple copy.

**Comment by Learath2** [comment]
On the issue of performance: Loading zips once at startup isn't that bad, we can bench it and if the few extra seconds is really that bad we can cache the extracted files.

On the issue of file-format: zip technically isn't the best tool for this job, I think we'd want something uncompressed so `tar` is more appropriate here, however the fact that zip is well supported on macOS and Windows makes it the better choice. I think under no circumstance should we ever roll our own, these binary file formats are disgustingly annoying to work with.

On the issue of gfx internals: I think @Jupeyy mentioned that single files would be fine and an atlas isn't needed. If an atlas is needed I think on VK and modern GL we can build it directly in the vram as we load. For the GL1.1 backend we can build the atlas on the CPU if it's not possible.

**Comment by heinrich5991** [comment]
> I think we'd want something uncompressed so `tar` is more appropriate here

I agree that it would be more appropriate here. Do note, however, that zip also supports storing files uncompressed.

**Comment by C0D3D3V** [comment]
Does it really matter if the zip is compressed or not? I thought we would open the zip, read all images to memory and close it again?

**Comment by heinrich5991** [comment]
> Does it really matter if the zip is compressed or not?

The compression is the thing that could cause the zip file to be slower than individual files.

**Comment by heinrich5991** [comment]
I think https://libzip.org/ might be a good choice for reading from zip files. It's 127 KiB on my machine.

**Comment by VoxelDoesCode** [comment]
I'm editing the game.png just a TINY bit to account for tiling in a better way, and I'm just wondering:
Would it be better for the gap between the hook chains to be centered (top) or is it better to keep it unbalanced (bottom)?
![image](https://github.com/ddnet/ddnet/assets/95713843/40d05076-388d-47a8-933a-3456c5e621a6)

**Comment by heinrich5991** [comment]
In what way is it unbalanced on the bottom? The right chain going in less than the left one?

> I'm editing the game.png just a TINY bit to account for tiling in a better way, and I'm just wondering:

It'd probably be best to split changes to the design from changes to the format, i.e. probably best to keep it "bad" for whatever value "bad" is and then fix it in a follow-up PR.

**Comment by VoxelDoesCode** [comment]
I mean quite literally, if you flip the hook chain asset it's not centered properly
![image](https://github.com/ddnet/ddnet/assets/95713843/c29d71c8-7ae1-435b-b74e-d764d643f55d)
I can worry about this later, the splitting's the easy part.

**Comment by VoxelDoesCode** [comment]
Another question: If we were to implement ZIP reading, how would we treat the actual ZIP *files*? Are we just going to have the same exact organization as before or are we going to be packaging EVERY needed image into a single ZIP? Because if we're going to be using the same organization, I don't know why we would. `game.png` is very unorganized, and it's pretty much just 5 different categories in one image. I don't know why we would want to continue this mess of an organization when we have the perfect opportunity to clean things up right now.

**Comment by heinrich5991** [comment]
You mean whether we'd keep the same game/hud/effects distinction? Sounds good to me. I'm not sure whether it should be in separate zips or not.

Do creators usually modify all of game.png/effects.png, etc. at once? If so, it'd make sense to combine all of them into one zip I guess.

**Comment by VoxelDoesCode** [comment]
I made a proof of concept [DDNet Image slicer](https://github.com/VoxelDoesCode/ddnet-image-slicer/tree/main), just to boost motivation of this project. Not sure if this way of handling the folders and ZIP files will be final, but we should discuss it before implementing it into the game.

**Comment by VoxelDoesCode** [comment]
This has come to my mind again, and it's gotten me thinking: Should we use an already established library for this? I've tried searching up ways to do it, but I have no idea how we would read the contents inside a ZIP file, but have it still be easy to maintain.

Another thing is that I think it would be nice to port the image slicer / image compiler packaged with the DDNet code, so you don't have to go to another GitHub page to do it. It should be as simple as running a .exe command.

Also, I think it'd be nice if we somehow made our own custom suffix for the filetype. I was thinking .twtp (meaning Teeworlds Texture Pack). I say this because a LOT of common files are really just ZIP files disguised. This begs the question on how every other program seemingly opens the contents, and reads from it. Does it secretly unzip it somewhere in a hidden file?

**Comment by heinrich5991** [comment]
As https://github.com/ddnet/ddnet/issues/6753#issuecomment-1649847738 says, you can use a library for reading zip files.

**Comment by VoxelDoesCode** [comment]
Oh, I didn't see that. It's been a while. I MIGHT try to start this up if no one else is doing so. I have high hopes that going this direction is going to aid DDNet massively.
